src/gtc/passes/oir_optimizations/inlining.py

```diff
@@ -0,0 +1,130 @@
+
+class MaskInlining(NodeTranslator):
+    """Inlines mask statements that are boolean mask fields with the expression that generates the field.
+
+    Preconditions: Mask statements exist as boolean temporary field accesses.
+    Postcondition: Mask statements exist as boolean expressions.
+    """
+
+    def visit_FieldAccess(
+        self,
+        node: oir.FieldAccess,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.Expr:
+        if node.name in masks_to_inline:
+            return cp.deepcopy(masks_to_inline[node.name])
+        return self.generic_visit(node, masks_to_inline=masks_to_inline, **kwargs)
+
+    def visit_AssignStmt(
+        self,
+        node: oir.AssignStmt,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.AssignStmt:
+        if node.left.name in masks_to_inline:
+            return None
+        return self.generic_visit(node, masks_to_inline=masks_to_inline, **kwargs)
+
+    def visit_MaskStmt(
+        self,
+        node: oir.MaskStmt,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.MaskStmt:
+        return oir.MaskStmt(
+            mask=self.visit(node.mask, masks_to_inline=masks_to_inline, **kwargs),
+            body=self.visit(node.body, masks_to_inline=masks_to_inline, **kwargs),
+        )
+
+    def visit_HorizontalExecution(
+        self,
+        node: oir.HorizontalExecution,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.HorizontalExecution:
+        return oir.HorizontalExecution(
+            body=[
+                stmt
+                for stmt in self.visit(
+                    node.body,
+                    masks_to_inline=masks_to_inline,
+                    **kwargs,
+                )
+                if stmt is not None
+            ],
+            declarations=node.declarations,
+        )
+
+    def visit_VerticalLoop(
+        self,
+        node: oir.VerticalLoop,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.VerticalLoop:
+        return oir.VerticalLoop(
+            loop_order=node.loop_order,
+            sections=self.visit(
+                node.sections,
+                masks_to_inline=masks_to_inline,
+                **kwargs,
+            ),
+            caches=[cache for cache in node.caches if cache.name not in masks_to_inline],
+        )
+
+    def visit_Stencil(self, node: oir.Stencil, **kwargs: Any) -> oir.Stencil:
+        masks_to_inline = MaskCollector().visit(node)
+        return oir.Stencil(
+            name=node.name,
+            params=node.params,
+            vertical_loops=self.visit(
+                node.vertical_loops,
+                masks_to_inline=masks_to_inline,
+                **kwargs,
+            ),
+            declarations=[decl for decl in node.declarations if decl.name not in masks_to_inline],
+        )
```

Comment:
This can be simplified to the following
```suggestion
    def visit_Temporary(self, node: oir.Temporary, *, masks_to_inline, **kwargs):
        return node if node.name not in masks_to_inline else NOTHING

    def visit_CacheDesc(self, node: oir.CacheDesc, *, masks_to_inline, **kwargs):
        return node if node.name not in masks_to_inline else NOTHING

    def visit_Stencil(self, node: oir.Stencil, **kwargs):
        return self.generic_visit(node, masks_to_inline=MaskCollector().visit(node), **kwargs)
```
Instead of removing temporaries and caches in the parent node, we can let them remove itself to get rid of some boiler-plate.